jib-core/src/main/java/com/google/cloud/tools/jib/image/ReproducibleLayerBuilder.java

```diff
@@ -100,9 +101,16 @@ public Blob build() {
     for (LayerEntry layerEntry : layerEntries) {
       // Adds the entries to uniqueTarArchiveEntries, which makes sure all entries are unique and
       // adds parent directories for each extraction path.
-      uniqueTarArchiveEntries.add(
+      TarArchiveEntry entry =
           new TarArchiveEntry(
-              layerEntry.getSourceFile().toFile(), layerEntry.getAbsoluteExtractionPathString()));
+              layerEntry.getSourceFile().toFile(), layerEntry.getAbsoluteExtractionPathString());
+
+      // Sets the entry's permissions by masking out the permission bits from the entry's mode (the
+      // lowest 9 bits) then using a bitwise OR to set them to the layerEntry's permissions.
+      entry.setMode(
+          (entry.getMode() & ~0777) | PermissionsHelper.toInt(layerEntry.getPermissions()));
```

Comment:
Have you seen the upper bits being used?

Reply:
They're used to specify file type (e.g. in `TarArchiveEntry`, files start with `0100` and directories start with `040`)